src/elexsolver/TransitionMatrixSolver.py

```diff
@@ -1,28 +1,107 @@
+        """
+        super().__init__()
+        self._strict = strict
+        self._lambda = lam
 
     @staticmethod
-    def __get_constraint(X, strict):
+    def __get_constraints(coef: np.ndarray, strict: bool) -> list:
         if strict:
-            return [cp.sum(X, axis=1) == 1]
-        return [cp.sum(X, axis=1) <= 1.1, cp.sum(X, axis=1) >= 0.9]
-
-    def __solve(self, A, B, strict):
-        transition_matrix = cp.Variable((A.shape[1], B.shape[1]))
-        loss_function = cp.norm(A @ transition_matrix - B, "fro")
-        objective = cp.Minimize(loss_function)
-        constraint = TransitionMatrixSolver.__get_constraint(transition_matrix, strict)
-        problem = cp.Problem(objective, constraint)
-        problem.solve()
+            return [0 <= coef, coef <= 1, cp.sum(coef, axis=1) == 1]
+        return [cp.sum(coef, axis=1) <= 1.1, cp.sum(coef, axis=1) >= 0.9]
+
+    def __standard_objective(self, A: np.ndarray, B: np.ndarray, beta: np.ndarray) -> cp.Minimize:
+        loss_function = cp.norm(A @ beta - B, "fro")
+        return cp.Minimize(loss_function)
+
+    def __ridge_objective(self, A: np.ndarray, B: np.ndarray, beta: np.ndarray) -> cp.Minimize:
+        # Based on https://www.cvxpy.org/examples/machine_learning/ridge_regression.html
+        lam = cp.Parameter(nonneg=True, value=self._lambda)
+        loss_function = cp.pnorm(A @ beta - B, p=2) ** 2
+        regularizer = cp.pnorm(beta, p=2) ** 2
+        return cp.Minimize(loss_function + lam * regularizer)
+
+    def __solve(self, A: np.ndarray, B: np.ndarray, weights: np.ndarray) -> np.ndarray:
+        transition_matrix = cp.Variable((A.shape[1], B.shape[1]), pos=True)
+        Aw = np.dot(weights, A)
+        Bw = np.dot(weights, B)
+
+        if self._lambda is None or self._lambda == 0:
+            objective = self.__standard_objective(Aw, Bw, transition_matrix)
+        else:
+            objective = self.__ridge_objective(Aw, Bw, transition_matrix)
+
+        constraints = TransitionMatrixSolver.__get_constraints(transition_matrix, self._strict)
+        problem = cp.Problem(objective, constraints)
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            try:
+                problem.solve(solver=cp.CLARABEL)
+            except (UserWarning, cp.error.SolverError) as e:
+                raise RuntimeError(e) from e
+
         return transition_matrix.value
 
-    def fit(self, A, B, strict=False):
-        transition_matrix = self.__solve(A, B, strict)
-        self.transition_matrix = transition_matrix
+    def fit_predict(self, X: np.ndarray, Y: np.ndarray, weights: np.ndarray | None = None) -> np.ndarray:
+        self._check_data_type(X)
+        self._check_data_type(Y)
+        self._check_any_element_nan_or_inf(X)
+        self._check_any_element_nan_or_inf(Y)
+
+        # matrices should be (units x things), where the number of units is > the number of things
+        if X.shape[1] > X.shape[0]:
```

Comment:
I'm not sure we should force this to be true. Not really true for our use-case but the optimization should still work if p > n

Reply:
It will, it will just transpose `X` (and then `Y` right below this), although I realize that might not be what everyone wants to happen.  I added in logic like this because I was getting really confused about the expected dimensions of the data and getting a lot of errors, so I added in code like this to standardize everything one way (units x things (candidates)).

The other issue is, if p > n, your result is going to be horrible if the solver even succeeds 😬  However, this might be something that's specific to voterflow and possibly other, but not all, uses of this solver, so I can move `_check_dimensions()` out of here and into `elex-voterflow-model`. 👍🏻 

But I guess even this check and transpose here on line 79 might also be specific to the use of the solver and not something inherently important to the solver itself 🤔

Ok, I'll try moving this to `elex-voterflow-model` as well and see what happens here 😬 😄